### ISSUE TITLE
Cross-reference take most of time when building large package with a probable bug in make_xref

### ISSUE THREAD

**Issue by junshi356** [sphinx-doc/sphinx#11282]
### Describe the bug

Building large packages takes a lot of time, most of which is cost on resolving cross-references. 

When building a package of 70MB and 3399 files ([azure-mgmt-containerservice](https://pypi.org/project/azure-mgmt-containerservice/)), it cost 3615 seconds and 2806 seconds are cost on resolving cross reference.
![image](https://user-images.githubusercontent.com/40595015/229457468-7755765b-b72c-4255-a690-399f7e8c075d.png)

"Fuzzy" searching mode of [PythonDomain.find_obj](https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L1346) method is iterating all loaded objects in the memory, when the package has a lot of objects and xref to resolve, the build process is very slow.

The `searchmode` parameter is designed to control whether using Fuzzy mode to resolve cross references, but I think it's not working under most cases (e.g. if I cross-reference the built-in `:class: str`, Sphinx will use fuzzy mode to match it). 

I think the problem is in [PyXrefMixin.make_xref](https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L377), attribute 'refspecific' is set to True, and won't be changed whatever the reftarget string is. Is it the expected behavior? Update the node attributes according to "refspecific" in return value back seems more reasonable after [parsing reftarget](https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L381).

And can we have a switch in config files to only find exact match and don't fuzzy search when building a package?

Screenshot:
Full iterating for fuzzy mode:
![image](https://user-images.githubusercontent.com/40595015/229457756-1d0c1aa5-a931-4864-bd26-1f8b4fdd33c1.png)

Seems should update the node attributes "refspecific" after `parse_reftarget`.
![image](https://user-images.githubusercontent.com/40595015/229469075-d20d50a7-88f8-4968-9300-109972b20858.png)

### How to Reproduce

Reproduce python script:
```
from sphinx.application import Sphinx

app = Sphinx(
        srcdir= './rst',
        confdir='./conf',
        outdir= './output',
        buildername='html',
        doctreedir='./doctrees'
    )
app.build()
```

Command to run cprofile:
```
 python -m cProfile -o "stats-compare" -s 'cumulative' .\run-sphinx-with-cprofile.py
```

RSTs are generated by apidoc: 
```
sphinx-apidoc --module-first --no-headings --no-toc --implicit-namespaces -o "" ""
```

conf.py:
```
# -*- coding: utf-8 -*-

source_suffix = '.rst'
master_doc = 'index'
project = u''
copyright = u''
author = u''
version = '0.1'
release = '0.1'
language = None
exclude_patterns = ['_build']
pygments_style = 'sphinx'
todo_include_todos = False
html_theme = 'bizstyle'
html_static_path = ['_static']
htmlhelp_basename = 'Example Document'
extensions = [
    'sphinx.ext.autodoc',
    'sphinx.ext.napoleon'
]

intersphinx_mapping = {'python': ('https://docs.python.org/3.6', None)}
# Make Google-style and Numpy-style Example work.
napoleon_use_admonition_for_examples = True

exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
autodoc_mock_imports = ['google']
autoclass_content = 'both'
autodoc_default_options = {
    'inherited-members': True
}
```

### Environment Information

```text
Platform:              win32; (Windows-10-10.0.22621-SP0)
Python version:        3.9.13 (tags/v3.9.13:6de2ca5, May 17 2022, 16:36:42) [MSC v.1929 64 bit (AMD64)])
Python implementation: CPython
Sphinx version:        6.1.3
Docutils version:      0.19
Jinja2 version:        3.0.1
Pygments version:      2.14.0
```

### Sphinx extensions

```python
['sphinx.ext.autodoc']
```

**Comment by picnixz** [comment]
After some investigation, here are my findings (feel free to argue):

`PyXrefMixin` always set `refspecific=True` and does not care about how the reference was presented. Consider the following example:

```rst

.. py:function:: foo(x)
   
   docstring 

   :param x: SOME_TYPE
   :type x: SOME_TYPE
```

For me, the value of `refspecific` should depend on  the configuration value `python_use_unqualified_type_names` and whether `.` is used as a prefix of `SOME_TYPE`. The presence of the `.` dictates the value of `refspecific`.

Now, I observed the following weird stuff in `PyXrefMixin.make_xref`. First, it is not possible to have `SOME_TYPE` equals to `.Bar` or `.foo.Bar` since otherwise the `reftarget` value is `.foo.Bar`. This is quite problematic and I would expect that the RHS of `:type ...:` to behave *as if* it were using <code>:class:&#96;SOME_TYPE&#96;</code>, as explained by:

https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L357-L359

The reason why I cannot use `:type x: .Bar` or `:type x: .foo.Bar` lies in:

https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L381-L395

<h4>Case <code>reftitle != reftarget</code></h4>

In this case, the `pending_xref` node will consist of a single text. This happens when prefixing the reference by `~` or `typing.` (e.g., `~foo.Bar` or `typing.Dict`). In other words, the target being resolved is everything after `~`. 

I don't know whether this is the proper usage of `~`, but when I use `~`, I use a fully qualified name afterwards. So, I don't see why we need to search with `refspecific=True`. Also, the way `parse_reftarget` is implemented forbids the combination of `~` with `.` It is possible to use <code>:class:&#96;~.foo.Bar&#96;</code> for instance, but when using it in <code>:type x: ~.foo.Bar</code>, we end up with a target `.foo.Bar` instead of `foo.Bar`.

<h4>Case <code>reftitle == reftarget</code> and <code>python_use_unqualified_type_names=True</code></h4>

In this case, the `pending_xref` node has two children and `reftarget == reftitle`. These two children are both required but their content should be pruned from any leading `.` if any. Again, one cannot use a type starting with a dot since it, e.g.:

```xml
<pending_xref py:class="True" py:module="main" refdomain="py" refexplicit="False" refspecific="True" reftarget=".foo.Bar" reftype="class">
  <pending_xref_condition condition="resolved">
    <literal_emphasis>
      Bar
  <pending_xref_condition condition="*">
    <literal_emphasis>
      .foo.Bar
```

Ideally, the reference target should be `foo.Bar` and the second resolution name should be `foo.Bar` and not `.foo.Bar`.

---

The optimizations I had in mind are quite simple. Everything is due to the fact that `refspecific=True` is assumed. This probably helps in many situations since that way the user does not need to put `.` explicitly. Instead, I suggest the following:

- Add a configuration value for the default value of `refspecific` in `PyXrefMixin`. For compatibility, it will be set to `True`. It can be `False`, in which case the user needs to put `.` or a fully qualified name (but explicit links are not supported, namely I cannot write `:type x: title <target>`, but this is a separate issue).

  If the value is `None`, we could internally decide the `refspecific` value. For instance, if the RHS of `:type x:` already contains dots (but does not start with one), then we could assume that it is sufficiently precise and won't check for all targets ending by that name. We should discuss about this approach as well. 

- Currently, the matches are constructed as follows:
 
  https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L1346-L1348

  Instead of indexing the objects by their name and iterating over the whole set of objects, I suggest sorting the dictionary in reverse alphabetical order so that the suffix can be matched on a range rather than iterating over the full collection of objects. 

  This approach however requires to freeze the dictionary or to implement a structure similar to use a list-like object for "insert & sort" operations but that can be used as a dictionary (for compatibility reasons). We could use https://grantjenks.com/docs/sortedcollections/ or a custom implementation. 

---

**Important note**

I was writing my reply when I wanted to check the list of object types being tested. And actually I think there is an important bug in `sphinx.ext.intersphinx`. Consider the following file:

```rst
.. py:function:: foo(x, y)

   docstring

   :param x: x
   :type x: typing.Any
   :param y: y
   :type y: typing.Any
```

In 

https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L1328-L1331

we have `objtypes = self.objtypes_for_role(type)`. But I was surprised to see the following output when `sphinx.ext.intersphinx` is loaded:

```python
# when processing x
['function', 'data', 'class', 'exception', 'method', 'classmethod', 'staticmethod', 'attribute', 'property', 'module']

# when processing y
['function', 'data', 'class', 'exception', 'method', 'classmethod', 'staticmethod', 'attribute', 'property', 'module', 'method'] 
```

In particular, there is some side-effect due to the dispatcher roles ! the `'method'` is duplicated ! and actually, it grows indefinitely. I confirmed that the bug only occurs if `sphinx.ext.intersphinx` is loaded:

https://github.com/sphinx-doc/sphinx/blob/5a6b2b16b6b588cdd4dab1b43b7873153da10baa/sphinx/ext/intersphinx.py#L365-L370

The `if` statements never check that the object types already exist. This is critical for performances. For small projects, we don't really care but for very large projects, it might be a real issue (**especially since there is an undesirable side-effect on the object types of the domain itself**).

**Comment by junshi356** [comment]
Thanks @picnixz ! Agree with you. 
I think another place to fix is not to assign `searchmode=1` when node has attribute of `refspecific` but the value is false. 
e.g. currently when `node['refspecific'] == false`, `searchmode = 1 if node.hasattr('refspecific') else 0` will still assign 1 to `searchmode`


https://github.com/sphinx-doc/sphinx/blob/39fa276b360dc36953b3ba62e19d6a978cd5109c/sphinx/domains/python.py#L1367-L1376

**Comment by picnixz** [comment]
>  I think another place to fix is not to assign searchmode=1 when node has attribute of refspecific but the value is false.

I resisted against that for this PR. I am not entirely sure that the Sphinx implementation actually assumes this or not. This is a trade-off between usability and time-space complexity (time because `has()` is faster than `get()` and space because it may save a `refspecific=False`). Now, *if* we allow a user specifying `refspecific` explicitly, we should indeed refactor this (possibly adding a DeprecationWarning along the way because it is quite important for extensions?).

By the way, `docutils` or `Sphinx` have the convention (at least for directives) to use a directive presence as an enabled flag. If we follow that convention (but don't allow custom default `refspecific`), this is not really a bug, but this should be documented somewhere (I don't know if it is already the case).

**Comment by junshi356** [comment]
OK, a config to turn-off default `refspecific` would be enough for unblocking us. And seems for the long term, the solution is to only assign `refspecific` when necessary.

**Comment by picnixz** [comment]
The duplication issue has been solved separately (#11337) but the reason why we are using a list instead of a set is to retain the discovery order. IIRC, the `objtypes` are initially ordered in priority order (and extensions could make changes to that). In intersphinx, we are iterating over that list so we need a structure that is order preserving *and* that has a O(1) lookup (amortized O(1) is also fine I think).
